File: ios/VisioMobile/Views/ParticipantListSheet.swift
import SwiftUI
import visioFFI

struct ParticipantListSheet: View {
    @EnvironmentObject private var manager: VisioManager
    @Environment(\.dismiss) private var dismiss

    private var lang: String { manager.currentLang }
    private var isDark: Bool { manager.currentTheme == "dark" }

    var body: some View {
        NavigationStack {
            List {
                // Local participant (always first)
                localParticipantRow

                // Remote participants sorted: hand raised first, then alphabetical
                let sorted = sortedParticipants
                ForEach(sorted, id: \.sid) { p in
                    participantRow(p)
                }
            }
            .scrollContentBackground(.hidden)
            .background(VisioColors.background(dark: isDark))
            .navigationTitle("\(Strings.t("participants.title", lang: lang)) (\(manager.participants.count + 1))")
            .navigationBarTitleDisplayMode(.inline)
            .toolbarColorScheme(isDark ? .dark : .light, for: .navigationBar)
            .toolbarBackground(VisioColors.surface(dark: isDark), for: .navigationBar)
            .toolbarBackground(.visible, for: .navigationBar)
            .toolbar(content: {
                ToolbarItem(placement: .topBarTrailing) {
                    Button {
                        dismiss()
                    } label: {
                        Image(systemName: "xmark")
                            .foregroundStyle(VisioColors.onSurface(dark: isDark))
                    }
                }
            })
        }
    }

    // MARK: - Local Participant Row

    private var localParticipantRow: some View {
        let name = manager.displayName.isEmpty
            ? Strings.t("call.you", lang: lang)
            : manager.displayName

        return HStack(spacing: 12) {
            avatarCircle(name: name)

            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 4) {
                    Text(name)
                        .font(.body)
                        .foregroundStyle(VisioColors.onSurface(dark: isDark))
                        .lineLimit(1)
                    Text("(\(Strings.t("call.you", lang: lang)))")
                        .font(.caption)
                        .foregroundStyle(VisioColors.secondaryText(dark: isDark))
                }
            }

            Spacer()

            statusIcons(
                isMuted: !manager.isMicEnabled,
                hasVideo: manager.isCameraEnabled,
                handRaisePosition: manager.isHandRaised ? 1 : 0,
                quality: .excellent
            )
        }
        .listRowBackground(VisioColors.surface(dark: isDark))
    }

    // MARK: - Remote Participant Row

    private func participantRow(_ p: ParticipantInfo) -> some View {
        let name = p.name ?? p.identity

        return HStack(spacing: 12) {
            avatarCircle(name: name)

            Text(name)
                .font(.body)
                .foregroundStyle(VisioColors.onSurface(dark: isDark))
                .lineLimit(1)

            Spacer()

            statusIcons(
                isMuted: p.isMuted,
                hasVideo: p.hasVideo,
                handRaisePosition: manager.handRaisedMap[p.sid] ?? 0,
                quality: p.connectionQuality
            )
        }
        .listRowBackground(VisioColors.surface(dark: isDark))
    }

    // MARK: - Shared Components

    private func avatarCircle(name: String) -> some View {
        let initials: String = {
            let parts = name.split(separator: " ")
            if parts.count >= 2 {
                return String(parts[0].prefix(1) + parts[1].prefix(1)).uppercased()
            }
            return String(name.prefix(2)).uppercased()
        }()

        let hue = Double(name.unicodeScalars.reduce(0) { $0 + Int($1.value) } % 360) / 360.0

        return ZStack {
            Circle()
                .fill(Color(hue: hue, saturation: 0.5, brightness: 0.35))
                .frame(width: 40, height: 40)
            Text(initials)
                .font(.system(size: 16, weight: .bold))
                .foregroundStyle(.white)
        }
    }

    private func statusIcons(
        isMuted: Bool,
        hasVideo: Bool,
        handRaisePosition: Int,
        quality: ConnectionQuality
    ) -> some View {
        HStack(spacing: 6) {
            if isMuted {
                Image(systemName: "mic.slash.fill")
                    .font(.system(size: 14))
                    .foregroundStyle(VisioColors.error500)
            }

            if !hasVideo {
                Image(systemName: "video.slash.fill")
                    .font(.system(size: 14))
                    .foregroundStyle(VisioColors.error500)
            }

            if handRaisePosition > 0 {
                HStack(spacing: 2) {
                    Image(systemName: "hand.raised.fill")
                        .font(.system(size: 11))
                    Text("\(handRaisePosition)")
                        .font(.caption2)
                        .bold()
                }
                .padding(.horizontal, 5)
                .padding(.vertical, 2)
                .background(VisioColors.handRaise)
                .clipShape(Capsule())
                .foregroundStyle(.black)
            }

            qualityIndicator(quality)
        }
    }

    @ViewBuilder
    private func qualityIndicator(_ quality: ConnectionQuality) -> some View {
        switch quality {
        case .excellent:
            Image(systemName: "wifi")
                .font(.system(size: 10))
                .foregroundStyle(.green)
        case .good:
            Image(systemName: "wifi")
                .font(.system(size: 10))
                .foregroundStyle(.yellow)
        case .poor:
            Image(systemName: "wifi.exclamationmark")
                .font(.system(size: 10))
                .foregroundStyle(.orange)
        case .lost:
            Image(systemName: "wifi.slash")
                .font(.system(size: 10))
                .foregroundStyle(VisioColors.error500)
        }
    }

    // MARK: - Sorting

    private var sortedParticipants: [ParticipantInfo] {
        manager.participants.sorted { a, b in
            let aRaised = manager.handRaisedMap[a.sid]
            let bRaised = manager.handRaisedMap[b.sid]

            // Hand raised first
            if aRaised != nil && bRaised == nil { return true }
            if aRaised == nil && bRaised != nil { return false }
            if let ap = aRaised, let bp = bRaised {
                if ap != bp { return ap < bp }
            }

            // Then alphabetical
            let aName = (a.name ?? a.identity).lowercased()
            let bName = (b.name ?? b.identity).lowercased()
            return aName < bName
        }
    }
}

File: android/app/src/main/kotlin/io/visio/mobile/ui/ParticipantListSheet.kt
package io.visio.mobile.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.visio.mobile.R
import io.visio.mobile.VisioManager
import io.visio.mobile.ui.i18n.Strings
import io.visio.mobile.ui.theme.VisioColors
import uniffi.visio.ParticipantInfo
import kotlin.math.absoluteValue

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ParticipantListSheet(
    participants: List<ParticipantInfo>,
    localDisplayName: String,
    localMicEnabled: Boolean,
    localCameraEnabled: Boolean,
    localIsHandRaised: Boolean,
    handRaisedMap: Map<String, Int>,
    lang: String,
    onDismiss: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)

    val totalCount = participants.size

    // The first participant is the local user (prepended by Rust).
    // Sort remaining participants: hand raised first, then alphabetically.
    val localP = participants.firstOrNull()
    val remoteParticipants = participants.drop(1)
    val sorted =
        remoteParticipants.sortedWith(
            compareByDescending<ParticipantInfo> { handRaisedMap[it.sid] != null }
                .thenBy { handRaisedMap[it.sid] ?: Int.MAX_VALUE }
                .thenBy { (it.name ?: it.identity).lowercase() },
        )

    val waitingParticipants by VisioManager.waitingParticipants.collectAsState()

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = VisioColors.PrimaryDark75,
    ) {
        // Header
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = "${Strings.t("participants.title", lang)} ($totalCount)",
                color = VisioColors.White,
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
            )
            IconButton(onClick = onDismiss, modifier = Modifier.size(32.dp)) {
                Icon(
                    painter = painterResource(R.drawable.ri_close_line),
                    contentDescription = Strings.t("participants.close", lang),
                    tint = VisioColors.White,
                    modifier = Modifier.size(20.dp),
                )
            }
        }

        // Waiting room section (host only — visible when there are waiting participants)
        if (waitingParticipants.isNotEmpty()) {
            Text(
                text = "${Strings.t("lobby.waitingParticipants", lang)} (${waitingParticipants.size})",
                style = MaterialTheme.typography.titleSmall,
                color = VisioColors.White,
                modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
            )
            waitingParticipants.forEach { participant ->
                Row(
                    modifier =
                        Modifier
                            .fillMaxWidth()
                            .padding(horizontal = 16.dp, vertical = 4.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(
                        text = participant.username,
                        color = VisioColors.White,
                        style = MaterialTheme.typography.bodyMedium,
                        modifier = Modifier.weight(1f),
                    )
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        Button(
                            onClick = { VisioManager.admitParticipant(participant.id) },
                            colors =
                                ButtonDefaults.buttonColors(
                                    containerColor = VisioColors.Primary500,
                                ),
                        ) {
                            Text(Strings.t("lobby.admit", lang))
                        }
                        OutlinedButton(
                            onClick = { VisioManager.denyParticipant(participant.id) },
                        ) {
                            Text(Strings.t("lobby.deny", lang))
                        }
                    }
                }
            }
            Spacer(modifier = Modifier.height(8.dp))
        }

        // Participant list
        LazyColumn(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            // Local participant (always first)
            if (localP != null) {
                item(key = localP.sid) {
                    val name =
                        localDisplayName.ifBlank {
                            localP.name ?: Strings.t("call.you", lang)
                        }
                    ParticipantRow(
                        name = name,
                        suffix = "(${Strings.t("call.you", lang)})",
                        isMuted = !localMicEnabled,
                        hasVideo = localCameraEnabled,
                        handRaisePosition = if (localIsHandRaised) 1 else 0,
                        qualityName = "Excellent",
                    )
                }
            }

            // Remote participants
            items(sorted, key = { it.sid }) { p ->
                ParticipantRow(
                    name = p.name ?: p.identity,
                    suffix = null,
                    isMuted = p.isMuted,
                    hasVideo = p.hasVideo,
                    handRaisePosition = handRaisedMap[p.sid] ?: 0,
                    qualityName = p.connectionQuality.name,
                )
            }
        }

        Spacer(modifier = Modifier.height(32.dp))
    }
}

@Composable
private fun ParticipantRow(
    name: String,
    suffix: String?,
    isMuted: Boolean,
    hasVideo: Boolean,
    handRaisePosition: Int,
    qualityName: String,
) {
    val initials =
        name
            .split(" ")
            .mapNotNull { it.firstOrNull()?.uppercase() }
            .take(2)
            .joinToString("")
            .ifEmpty { "?" }

    val hue = name.fold(0) { acc, c -> acc + c.code }.absoluteValue % 360
    val avatarColor = Color.hsl(hue.toFloat(), 0.5f, 0.35f)

    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        // Avatar
        Box(
            modifier =
                Modifier
                    .size(40.dp)
                    .clip(CircleShape)
                    .background(avatarColor),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = initials,
                color = VisioColors.White,
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
            )
        }

        // Name + suffix
        Row(
            modifier = Modifier.weight(1f),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Text(
                text = name,
                color = VisioColors.White,
                fontSize = 15.sp,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f, fill = false),
            )
            if (suffix != null) {
                Text(
                    text = suffix,
                    color = VisioColors.Greyscale400,
                    fontSize = 13.sp,
                    maxLines = 1,
                )
            }
        }

        // Status icons
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            // Mic muted
            if (isMuted) {
                Icon(
                    painter = painterResource(R.drawable.ri_mic_off_fill),
                    contentDescription = null,
                    tint = VisioColors.Error500,
                    modifier = Modifier.size(16.dp),
                )
            }

            // Camera off
            if (!hasVideo) {
                Icon(
                    painter = painterResource(R.drawable.ri_video_off_fill),
                    contentDescription = null,
                    tint = VisioColors.Error500,
                    modifier = Modifier.size(16.dp),
                )
            }

            // Hand raised
            if (handRaisePosition > 0) {
                Row(
                    modifier =
                        Modifier
                            .background(VisioColors.HandRaise, RoundedCornerShape(10.dp))
                            .padding(horizontal = 5.dp, vertical = 1.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(2.dp),
                ) {
                    Icon(
                        painter = painterResource(R.drawable.ri_hand),
                        contentDescription = null,
                        tint = Color.Black,
                        modifier = Modifier.size(12.dp),
                    )
                    Text(
                        text = "$handRaisePosition",
                        color = Color.Black,
                        fontSize = 11.sp,
                        fontWeight = FontWeight.SemiBold,
                    )
                }
            }

            // Connection quality
            ConnectionQualityBarsSmall(qualityName)
        }
    }
}

@Composable
private fun ConnectionQualityBarsSmall(quality: String) {
    val bars =
        when (quality) {
            "Excellent" -> 3
            "Good" -> 2
            "Poor" -> 1
            else -> 0
        }
    Row(
        horizontalArrangement = Arrangement.spacedBy(1.dp),
        verticalAlignment = Alignment.Bottom,
    ) {
        for (i in 1..3) {
            Box(
                modifier =
                    Modifier
                        .width(3.dp)
                        .height((i * 3 + 2).dp)
                        .background(
                            if (i <= bars) Color.Green else VisioColors.Greyscale400,
                            RoundedCornerShape(1.dp),
                        ),
            )
        }
    }
}
